---
title: "Loading Deleteome and McCormick 2015 Survival Data"
author: "Will Townes"
date: "9/14/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#library(rhdf5) #from bioconductor
library(dplyr)
library(SingleCellExperiment) #also loads SummarizedExperiment
library(biomaRt)
library(Matrix)
#mutants<-"all"
mutants<-"responsive"
```

# Load and clean raw data

## GenAge Database

```
wget -P data https://genomics.senescence.info/genes/models_genes.zip
unzip data/models_genes.zip -d data
rm data/release.html
rm data/models_genes.zip
```

```{r}
d<-read.csv("data/genage_models.csv")
d1 <- d %>% subset(organism=="Saccharomyces cerevisiae" & longevity.influence %in% c("Pro-Longevity","Anti-Longevity")) %>% transform(symbol=toupper(symbol)) %>% mutate(pro_longevity=as.numeric(longevity.influence=="Pro-Longevity")) %>% select(symbol,pro_longevity)
#write.table(d1,file="data/genage_yeast.txt",row.names=FALSE)
```

## GO terms

```{r}
#spp<-c("scerevisiae","celegans","dmelanogaster","mmusculus")
f<-function(org,verbose=TRUE){
  if(verbose){ print(paste("species:",org)) }
  org_match<-list(yeast="scerevisiae",worm="celegans",fly="dmelanogaster",mouse="mmusculus")
  x<-useMart("ensembl",dataset=paste0(org_match[[org]],"_gene_ensembl"))
  x<-getBM(c("external_gene_name","go_id"),mart=x)
  x$species<-org
  x
}

res<-do.call("rbind",lapply(names(org_match),f))
table(res$species)/nrow(res) #fraction of entries from each species
write.table(res,file="./data/goterms.txt",row.names=FALSE)
```

## Worm Cell Atlas

Single cell RNA-Seq for worm larva. Documentation: http://atlas.gs.washington.edu/worm-rna/docs/

```
wget -P data http://waterston.gs.washington.edu/sci_RNA_seq_gene_count_data/Cao_et_al_2017_vignette.RData
```

```{r}
cds<-attach("./data/Cao_et_al_2017_vignette.RData")$cds
umi<-exprs(cds)
gm<-fData(cds)
cm<-pData(cds)
gc<-!is.na(cm$tissue) & !is.na(cm$cell.type) & cm$tissue != "Failed QC"
cm<-cm[gc,]
cm$tissue<-factor(cm$tissue)
cm$cell.type<-factor(cm$cell.type)
umi<-umi[,gc]
gg<-gm$num_cells_expressed>0
gm<-gm[gg,]
umi<-umi[gg,]
rownames(gm)<-rownames(umi)<-toupper(gm$symbol)
rab<-t(t(umi)/colSums(umi))
sce<-SingleCellExperiment(assays=list(umi_counts=umi,relative_abundance=rab),rowData=gm,colData=cm)
#sizeFactors(sce)<-cm$Size_Factor
#sce<-scater::normalize(sce,exprs_values="umi", return_log=TRUE, log_exprs_offset=1, centre_size_factors=FALSE)
saveRDS(sce,"./data/worm_scrna_atlas.rds")
#sce<-readRDS("./data/worm_scrna_atlas.rds")
umi<-assay(sce,"umi_counts")
cm<-colData(sce)
cm2<-data.frame(celltype=cm$cell.type,plate=cm$plate)
cm2$grp<-factor(paste0(cm2$celltype,"_p",cm2$plate))
f<-function(g){
  x<-as.matrix(umi[,cm2$grp==g])
  drop(rowSums(x))
}
umi2<-vapply(levels(cm2$grp),f,FUN.VALUE=rep(0,nrow(umi)))
write.table(umi2,"./data/worm_scrna_atlas_merged_umi.txt")
```

## ARCHS4

```
wget -P data https://s3.amazonaws.com/mssm-archs4-zoo/Caenorhabditis_elegans_genecount_v1.h5
wget -P data https://s3.amazonaws.com/mssm-archs4-zoo/Saccharomyces_cerevisiae_genecount_v1.h5
```

```{r}
get_gene_info<-function(g,org=c("worm","yeast"),verbose=TRUE){
  #g is a character vector of gene IDs from ARCHS4
  #this function maps IDS to gene symbols (compatible with genage)
  #also includes gene length in the returned data frame
  org<-match.arg(org)
  if(verbose){ print(paste("species:",org)) }
  org_match<-list(yeast=c("scerevisiae","ensembl_gene_id"),
                  worm=c("celegans","wormbase_gene"))
  #,fly="dmelanogaster",mouse="mmusculus")
  x<-useMart("ensembl",dataset=paste0(org_match[[org]][1],"_gene_ensembl"))
  gname<-org_match[[org]][2] #eg, "wormbase_gene"
  z<-getBM(c(gname,"external_gene_name","start_position","end_position"),filters=gname,values=g,mart=x)
  z$length<-z$end_position-z$start_position
  z$external_gene_name<-toupper(z$external_gene_name)
  z
}

load_archs4<-function(species=c("worm","yeast"),dir="./data",v="v1"){
  org_match<-list(yeast="Saccharomyces cerevisiae",worm="Caenorhabditis elegans")
                #,fly="Drosophila melanogaster",mouse="Mus musculus")
  sp<-org_match[[match.arg(species)]] #eg "Caenorhabditis elegans"
  fn<-paste(c(unlist(strsplit(sp," ")),"genecount",v),collapse="_")
  hdf5path<-file.path(dir,paste0(fn,".h5"))
  d<-rhdf5::h5read(hdf5path,"data/expression")
  m<-rhdf5::h5read(hdf5path,"meta")
  i<-rhdf5::h5read(hdf5path,"info")
  cmeta<-as.data.frame(do.call(cbind,m[names(m)!="genes"]))
  colnames(d)<-cmeta$Sample_geo_accession
  g<-as.character(m[["genes"]])
  rownames(d)<-g
  gmeta<-get_gene_info(g,species,verbose=FALSE)
  gmeta[,c("start_position","end_position")]<-NULL
  d<-d[gmeta[,1],]
  gg<-gmeta$external_gene_name!="" & !duplicated(gmeta$external_gene_name)
  gmeta<-gmeta[gg,]
  d<-d[gg,]
  rownames(d)<-rownames(gmeta)<-gmeta$external_gene_name
  SummarizedExperiment(list(counts=d),colData=cmeta,metadata=i,rowData=gmeta)
}
for(org in c("yeast","worm")){
  se<-load_archs4(org)
  saveRDS(se,file.path("./data",paste0("archs4",org,".rds")))
}
```

## McCormick 2015 Replicative Lifespan

Useful functions, mostly for McCormick data

```{r}
merge_lifespans<-function(x){
  #x is a vector like c("4,3,5","2,3,1","9")
  #returns a single string concatenating all values together
  paste0(x,collapse=",")
}
str2vec<-function(x){as.numeric(unlist(strsplit(x,",",fixed=TRUE)))}
str2mean<-function(y){
  vapply(y,function(t){mean(str2vec(t))},FUN.VALUE=1.0)
}
str2count<-function(y){
  vapply(y,function(t){length(str2vec(t))},FUN.VALUE=1)
}
ranksum_test_pval<-function(set_vals,ref_vals,exact=FALSE,...){
  wilcox.test(str2vec(set_vals),str2vec(ref_vals),exact=exact,...)$p.value
}
str2pval<-function(x,y){
  mapply(ranksum_test_pval,x,y)
}
perts2uniquegenes<-function(x){
  #x a list of space-separated gene symbols
  #each item is a perturbation, can have many genes per perturbation
  #returns a vector with how many genes were in each perturbation
  sg<-strsplit(as.character(x)," ",fixed=TRUE)
  mcp<-unique(unlist(sg))
  toupper(mcp) #list of unique gene symbols perturbed
}
ngenes_per_perturbation<-function(x){
  sg<-strsplit(as.character(x)," ",fixed=TRUE)
  sapply(sg,length)
}
```

This is genetic perturbation and replicative lifespan data for almost all genes in yeast. The original data was provided by Kareem Carr who had emailed the author and received an excel file in response. Also, the authors provided this information:

```
---------- Forwarded message ----------
From: Kareem Carr <kareemcarr@fas.harvard.edu>
Date: Sat, Jun 17, 2017 at 5:26 PM
Subject: Fwd: A Comprehensive Analysis of Replicative Lifespan ... [Data Request]
To: Jeff Miller <jwmiller@hsph.harvard.edu>, "Doshi-Velez, Finale" <finale@seas.harvard.edu>

Full lifespan data set from A Comprehensive Analysis of Replicative Lifespan in 4,698 Single-Gene Deletion Strains Uncovers Conserved Mechanisms of Aging by McCormick et al.

---------- Forwarded message ----------
From: Mark McCormick <mark.mccormick.01@gmail.com>
Date: Fri, Jun 16, 2017 at 6:29 PM
Subject: Re: A Comprehensive Analysis of Replicative Lifespan ... [Data Request]
To: Kareem Carr <kareemcarr@fas.harvard.edu>

Hi Kareem, here is a summary of our RLS data, for all BY background deletions (singly and in combination(s)), on YPD plates at 30C, as of the time we compiled the results for the screen.

This is by file (experiment), so if there are multiple rows for the same genotype in this file, these rows are all mutually exclusive.  This means that to get the complete data for any genotype you would combine the raw RLS data listed under "set_lifespans" for all rows sharing that genotype, and to get the corresponding control data you would combine all "ref_lifespans" cells listed under that genotype.

This is missing some corrupted data from a cryptolocker trojan that got onto someone's laptop, when a previous tech in the lab hadn't backed up the database versions correctly.   With this in mind, there is one particular class of data points that I'd point out, that you'll be able to spot:  if you compile all of the data for a given genotype, and the % change is above +30%, yet the n is only 5 cells, there is data missing for that genotype.  Specifically, we always added at least 20 more cells for anything that tested over +30% at the 5-cell n, to confirm whether it was long-lived at a higher n, and it was we added many more cells later on.  If you look at these cases, they are almost always examples where the deletion is around 26 or so (typical WT), and the WT control is unusually short lived, so these are genotypes that did not test out to be long lived once 20 cells were added, but where that subsequent not-long-lived 20 cell retest data was among the data lost to the cryptolocker fiasco.  Hope that makes sense.
```
Translation: exclude any perturbation genotype with percent_change>30 and set_lifespan_count<=5.

```{r}
d0<-read.csv("data/mccormick2015.csv",stringsAsFactors=FALSE)
d0$set_genotype<-toupper(d0$set_genotype)
#remove bad genotypes with missing data
#need to pool across experiments
d<-d0[,c("set_genotype","set_lifespans","ref_lifespans")]
d<-d %>% group_by(set_genotype) %>% summarise(set_lifespans=merge_lifespans(set_lifespans),ref_lifespans=merge_lifespans(ref_lifespans))
d<-d %>% transform(set_lifespan_count=str2count(set_lifespans),set_lifespan_mean=str2mean(set_lifespans),ref_lifespan_mean=str2mean(ref_lifespans))
d<-d %>% transform(percent_change=(set_lifespan_mean-ref_lifespan_mean)*100/ref_lifespan_mean)
bad<-d %>% filter(percent_change>30 & set_lifespan_count<=5)
bad<-bad$set_genotype
d<-d0[!(d0$set_genotype %in% bad),]

sort(table(d$ref_genotype),decreasing=TRUE)
#restrict data to only the reference genotypes BY4742 and BY4741 (this is almost everything)
d<-d[d$ref_genotype %in% c("BY4742","BY4741"),]
#restrict to experiments with same ref and set background
d<-d[d$ref_background==d$set_background,]
#misc data cleanup, remove extra cols
#all experiments had set media YPD and temp 30C
extra_cols<-c("ref_locus_tag","set_locus_tag","pooled_by","ref_temperature","set_temperature","set_media","ref_media")
d[,extra_cols]<-NULL
#how many single gene vs multi gene deletions?
#mcp<-perts2uniquegenes(d$set_genotype)
ng<-ngenes_per_perturbation(d$set_genotype)
table(ng)
d$numgenes<-ng

#merge all experiments from same genotype
d<-d %>% mutate(mean_lifespan_change=set_lifespan_mean-ref_lifespan_mean,ivar_wt=1/(1/set_lifespan_count+1/ref_lifespan_count))
d<-d %>% group_by(set_genotype) %>% summarise(lifespan_change=sum(mean_lifespan_change*ivar_wt)/sum(ivar_wt), set_lifespans=merge_lifespans(set_lifespans), ref_lifespans=merge_lifespans(ref_lifespans), max_pval=max(ranksum_p), min_pval=min(ranksum_p))
d<-d %>% mutate(combined_pval=str2pval(set_lifespans,ref_lifespans))
d<-d %>% select(-c(set_lifespans,ref_lifespans))
d<-d %>% mutate(numgenes=ngenes_per_perturbation(set_genotype))
d %>% subset(set_genotype=="LOS1") #should be about +7
write.csv(as.data.frame(d),file="data/mccormick_all.csv",row.names=FALSE)
```

## Kemmeren 2014 Gene Expression

Gene expression (2-color microarray) for about 1500 single gene deletion yeast mutants. The data were obtained from http://deleteome.holstegelab.nl. Click downloads and select all mutants, not excluding wild type variable genes.

```
wget -P data http://deleteome.holstegelab.nl/data/downloads/deleteome_all_mutants_controls.txt
wget -P data http://deleteome.holstegelab.nl/data/downloads/deleteome_responsive_mutants_ex_wt_var_controls.txt
```

```{r}
#Read in data
if(mutants=="responsive"){
  d <- read.delim("data/deleteome_responsive_mutants_ex_wt_var_controls.txt",stringsAsFactors=FALSE)
} else if(mutants=="all"){
  d <- read.delim("data/deleteome_all_mutants_controls.txt",stringsAsFactors=FALSE)
} else { stop("invalid mutant choice") }
df2mat<-function(df){
  #convert a string data frame into a numeric matrix of same dimension
  matrix(as.numeric(as.matrix(df)),nrow=nrow(df),dimnames=dimnames(df))
}
gdata<-d[2:nrow(d),1:3]
guniq<-!duplicated(gdata$geneSymbol)
gdata<-gdata[guniq,]
d_a<-d[2:nrow(d),seq(from=4,to=ncol(d),by=3)] #average of logs
d_m<-d[2:nrow(d),seq(from=5,to=ncol(d),by=3)] #difference of logs
d_p<-d[2:nrow(d),seq(from=6,to=ncol(d),by=3)] #p-values
#exclude the three control experiments at the end and duplicated genes
d_a<-df2mat(d_a[guniq,1:(ncol(d_a)-3)])
d_m<-df2mat(d_m[guniq,1:(ncol(d_m)-3)])
d_p<-df2mat(d_p[guniq,1:(ncol(d_p)-3)])
cnames<-colnames(d_a)
colnames(d_m)<-colnames(d_p)<-cnames
#extract symbol of gene that was perturbed
perts<-sapply(strsplit(cnames,"\\."),function(t){toupper(t[1])})
#how many perturbation genes were measured?
table(perts %in% gdata$geneSymbol)
cdata<-data.frame(pert_geneSymbol=perts)
rownames(cdata)<-cnames
se<-SummarizedExperiment(list(logratio=d_m,avglog=d_a,pval=d_p),rowData=gdata,colData=cdata,metadata=list("logratio: M values, avglog: A values, the expression values are a contrast between gene knockout perturbation versus control"))
rownames(se)<-gdata$geneSymbol
if(mutants=="all"){
  saveRDS(se,"data/deleteome_all_mutants_controls.rds")
} else if(mutants=="responsive") {
  saveRDS(se,"data/deleteome_responsive_mutants.rds")
}
```

### old stuff

split clusters in two during burn-in for Gibbs
use weak limit Dirichlet prior over class probabilities

```{r}
#old stuff
#sv<-readRDS("data/mccormick_all.rds")
#compare overlap between deleteome and mccormick
gdata<-rowData(ge)
"VMA2" %in% gdata$geneSymbol
#overlap between deleteome perts and survival perts
perts<-colData(ge)$pert_geneSymbol
table(perts %in% mcp)
table(perts %in% mcp,perts %in% gdata$geneSymbol)
#overlap between deleteome gene measurements and survival perts
mcp<-perts2uniquegenes(sv$set_genotype)
table(mcp %in% gdata$geneSymbol)
table(gdata$geneSymbol %in% mcp)
head(mcp[!(mcp %in% gdata$geneSymbol)])
#combine data for use in regression models
```