---
title: "Predicting Survival from Gene Expression"
author: "Will Townes"
date: "3/29/2018"
output: html_document
---

```{r}
library(SummarizedExperiment)
library(dplyr)
library(class)
library(glmnet)
library(caret)
library(ROCR)
mutants<-"responsive"
binarize<-TRUE
measured_genes<-TRUE
```

## Data Loading

Merge Gene Expression and Survival Data
Find genes that were measured in the deleteome across perturbations and also for which we have perturbation-survival data.

```{r}
sv<-read.csv("data/mccormick_all.csv",stringsAsFactors=FALSE)
sv<-subset(sv,numgenes==1)

if(mutants=="all"){
  ge<-readRDS("data/deleteome_all_mutants_controls.rds")
} else if(mutants=="responsive"){
  ge<-readRDS("data/deleteome_responsive_mutants.rds")
} else { stop("invalid mutant") }

if(measured_genes==TRUE){
  #expressed genes vs RLS perturbs
  table(rownames(ge) %in% sv$set_genotype)
  table(sv$set_genotype %in% rownames(ge))
  sv<-sv %>% subset(set_genotype %in% rownames(ge))
  ge<-ge[rownames(ge) %in% sv$set_genotype,]
  X<-assay(ge,"logratio")[sv$set_genotype,] #sort order match
} else {
  #perturbed genes vs RLS perturbs
  ge_perts<-colData(ge)$pert_geneSymbol
  table(ge_perts %in% sv$set_genotype)
  table(sv$set_genotype %in% ge_perts)
  sv<-sv %>% subset(set_genotype %in% ge_perts)
  ge<-ge[,ge_perts %in% sv$set_genotype]
  rownames(ge)<-paste0("expr_",rownames(ge))
  colnames(ge)<-colData(ge)$pert_geneSymbol
  X<-t(assay(ge,"logratio")[,sv$set_genotype])
}
if(binarize==TRUE){
  y<-as.numeric(sv$lifespan_change>0 & sv$combined_pval<.05)
} else {
  y<-sv$lifespan_change
}
names(y)<-rownames(X)

# train/test split
train_idx<-createDataPartition(y,p=.75,list=FALSE)
ytrn<-y[train_idx]
ytst<-y[-train_idx]
```

knn classifier

```{r}
Xtrn<-X[train_idx,]
Xtst<-X[-train_idx,]
fit<-knn(Xtrn,Xtst,ytrn,k=10,prob=TRUE)
probs<-attr(fit,"prob")
probs[fit==0]<-1-probs[fit==0]
preds<-prediction(probs,ytst)
auc.tmp<-performance(preds,"auc")
(auc<-as.numeric(auc.tmp@y.values)) #0.6
```

regression-type models

```{r}
pca<-FALSE
if(pca){
  Xs<-X %*% svd(X,nu=500,nv=500)$v #if PCA adjustment
  colnames(Xs)<-paste0("pca_",seq.int(500))
} else {
  #random subset of columns or use PCA to reduce dims
  nsample<-min(1000,ncol(X))
  Xs<-X[,sample.int(ncol(X),nsample)]
}
Xtrn<-Xs[train_idx,]
Xtst<-Xs[-train_idx,]
```

gradient boosted machines

```{r}
distrib<-if(binarize==TRUE) "bernoulli" else "gaussian"
system.time(res2<-train(Xtrn,factor(ytrn), method="gbm", distribution=distrib, preProcess=NULL, verbose=FALSE, metric="Kappa", trControl=trainControl("cv",5), tuneGrid=expand.grid(interaction.depth=c(2,3),n.trees=c(800,1600),shrinkage=.005,n.minobsinnode=10)))
print(res2)
probs<-predict(res2,newdata=Xtst,type="prob")[,"1"]
preds<-prediction(probs,ytst)
auc.tmp<-performance(preds,"auc")
(auc<-as.numeric(auc.tmp@y.values)) #0.6
```

svm

```{r}
distrib<-if(binarize==TRUE) "bernoulli" else "gaussian"
system.time(res3<-train(Xtrn,factor(ytrn), method="svmRadial", distribution=distrib, preProcess=NULL, verbose=FALSE, metric="Kappa", trControl=trainControl("cv",5)))
print(res3)
probs<-predict(res3,newdata=Xtst,type="prob")[,"1"]
preds<-prediction(probs,ytst)
auc.tmp<-performance(preds,"auc")
(auc<-as.numeric(auc.tmp@y.values)) #0.6
```

GLMnet prediction

```{r}
pp<-preProcess(Xtrn,method=c("center","scale"))
Xtrn<-cbind(1,predict(pp,Xtrn))
Xtst<-cbind(1,predict(pp,Xtst))

#folds<-createFolds(ytrn)
#res<-train(mean_lifespan~.,data=d[train_idx,],preProcess=c("center","scale"))

alpha<-if(pca){ 0.0 }else{ .9 }
if(binarize==TRUE){
  res<-cv.glmnet(Xtrn,ytrn,family="binomial",alpha=alpha)
} else {
  res<-cv.glmnet(Xtrn,ytrn,alpha=alpha)
}
probs<-predict(res, newx=Xtst, s="lambda.min", type="response")
#saveRDS(res,file="glmnet_fit.rds")
plot(res)
preds<-prediction(probs,ytst)
auc.tmp<-performance(preds,"auc")
(auc<-as.numeric(auc.tmp@y.values)) #0.5
```

```{r}
rmse<-function(ytrue,ypred){
  sqrt(mean((ytrue-ypred)^2))
}
(rmse_glmnet<-rmse(ypred,ytst))
(rmse_mean_only<-rmse(mean(ytrn),ytst))
```

